### PR TITLE
Introduce the uninstall command

### PR DESCRIPTION
**Pull request by NatashaWso2**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/10260

## Automation tests
 - Integration tests
   > Both positive and negative integration tests are added covering all the areas

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+/**
+ * This class provides util methods for uninstalling Ballerina packages.
+ *
+ * @since 0.982.0
+ */
+public class UninstallUtils {
+    private static final Path BALLERINA_HOME_PATH = RepoUtils.createAndGetHomeReposPath();
+    private static PrintStream outStream = System.out;
+
```
Comment (MaryamZi) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: extra space?

Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Removed the extra line with c1495d8

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+     * @param fullPkgPath package provided
+     */
+    public static void uninstallPackage(String fullPkgPath) {
+        String orgName;
+        String packageName;
+        String version;
+        // Get org-name
+        int orgNameIndex = fullPkgPath.indexOf("/");
+        if (orgNameIndex != -1) {
+            orgName = fullPkgPath.substring(0, orgNameIndex);
```
Comment (Kishanthan) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Lets fail fast. Inverse this if and throw the error fast.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        } else {
+            throw new BLangCompilerException("no org-name is provided");
+        }
+        // Get package name
+        int packageNameIndex = fullPkgPath.indexOf(":");
+        if (packageNameIndex != -1) { // version is provided
+            packageName = fullPkgPath.substring(orgNameIndex + 1, packageNameIndex);
+            version = fullPkgPath.substring(packageNameIndex + 1, fullPkgPath.length());
+        } else {
+            throw new BLangCompilerException("no package version is provided");
```
Comment (Kishanthan) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Same here also. Lets fail fast. Inverse this if and throw the error fast.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    /**
+     * Delete empty directories.
+     *
+     * @param from package directory path
+     * @param to   home repository path
+     * @throws IOException throw an exception if an error occurs
+     */
+    private static void deleteEmptyDirsUpTo(Path from, Path to) throws IOException {
```
Comment (Kishanthan) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: This method is not clear. Why do we need the second arg "to" ?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java (hunk 2 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Fixed with 26505f6

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java (hunk 3 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Fixed with 26505f6

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java (hunk 4 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: After deleting the package.zip from the home repository we should delete the empty directories as well. We need to delete all folders from the package path to the home directory only if they are empty (going backwards).

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+     * @param fullPkgPath package provided
+     */
+    public static void uninstallPackage(String fullPkgPath) {
+        String orgName;
+        String packageName;
+        String version;
+        // Get org-name
+        int orgNameIndex = fullPkgPath.indexOf("/");
+        if (orgNameIndex == -1) {
+            throw new BLangCompilerException("no org-name is provided");
```
Comment (hemikak) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Is BLangCompilerException the right type of exception here ?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        }
+        orgName = fullPkgPath.substring(0, orgNameIndex);
+
+        // Get package name
+        int packageNameIndex = fullPkgPath.indexOf(":");
+        if (packageNameIndex == -1) { // version is provided
+            throw new BLangCompilerException("no package version is provided");
+        }
+        packageName = fullPkgPath.substring(orgNameIndex + 1, packageNameIndex);
+        version = fullPkgPath.substring(packageNameIndex + 1, fullPkgPath.length());
```
Comment (hemikak) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: If version is not given, version will be taken as empty and continued. Don't we need to verify ? If we are verifying use the semver regex to check.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        Path pkgPath = Paths.get(orgName, packageName, version);
+
+        // Check if package is installed locally
+        if (Files.exists(homeRepoPath.resolve(pkgPath), LinkOption.NOFOLLOW_LINKS)) {
+            deletePackage(homeRepoPath, pkgPath, fullPkgPath, packageName);
+        } else if (Files.exists(cacheRepoPath.resolve(pkgPath), LinkOption.NOFOLLOW_LINKS)) {
+            deletePackage(cacheRepoPath, pkgPath, fullPkgPath, packageName);
+        } else {
+            // package to be uninstalled doesn't exists
+            throw new BLangCompilerException("package does not exist " + fullPkgPath);
```
Comment (hemikak) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Better if we can say to check the package signature in the exception.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        int orgNameIndex = fullPkgPath.indexOf("/");
+        if (orgNameIndex == -1) {
+            throw new BLangCompilerException("no org-name is provided");
+        }
+        orgName = fullPkgPath.substring(0, orgNameIndex);
+
+        // Get package name
+        int packageNameIndex = fullPkgPath.indexOf(":");
+        if (packageNameIndex == -1) { // version is provided
+            throw new BLangCompilerException("no package version is provided");
```
Comment (hemikak) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Is BLangCompilerException the right type of exception here ?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java (hunk 7 above)
Comment (hemikak) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Is BLangCompilerException the right type of exception here ?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        } else {
+            // package to be uninstalled doesn't exists
+            throw new BLangCompilerException("package does not exist " + fullPkgPath);
+        }
+    }
+
+    /**
+     * Remove package from home repository.
+     *
+     * @param repoPath    path of the home repository
```
Comment (hemikak) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Better to explain this param more.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+     * @param pkgPath     package path
```
Comment (hemikak) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: reword to 'package directory path'

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+     * @param fullPkgPath full package path as given by user
```
Comment (hemikak) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: This is more of a package ID isn't it ? But have we used package path across other implementations ?

Comment (hemikak) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Also mention that it is for logging purposes only.

Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: This is not the Package ID. fullPkgPath is the package to be uninstalled which is provided by the user. I will add a comment that it is used only for logging purposes.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java (hunk 5 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Changed the exception type with 1d90337

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java (hunk 6 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: The version is verified with https://github.com/NatashaWso2/ballerina/blob/uninstall-command/cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java#L58. If the version is not given we throw an error to the user

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java (hunk 7 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Changed the exception type and changed the message to "incorrect package signature provided" with 1d90337

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java (hunk 8 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Changed the exception type with 1d90337

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java (hunk 9 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Added a proper description with 1d90337

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java (hunk 10 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Changed parameter name to "pkgDirectoryPath" and updated the java docs with 1d90337

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java (hunk 4 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Changed the method name to  'deleteEmptyParentDirs' with e74db59

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/HomeDirRepositoryProvider.java [hunk 12]
```diff
@@ -1,26 +0,0 @@
...
-
-import static org.ballerinalang.util.BLangConstants.USER_REPO_ARTIFACTS_DIRNAME;
-import static org.ballerinalang.util.BLangConstants.USER_REPO_SRC_DIRNAME;
-
-/**
- * Load bal files form $HOME/.ballerina.
- * <p>
- * ie: import foo.bar; will load from ~/.ballerina/artifacts/src/foo/bar/*.bal
- */
-@JavaSPIService("org.ballerinalang.spi.UserRepositoryProvider")
```
Comment (hemikak) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/HomeDirRepositoryProvider.java]: Why dont we need the SPI ?

Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/HomeDirRepositoryProvider.java]: This SPI is not used anymore so deleted the file

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+     * @param fullPkgPath package provided
+     */
+    public static void uninstallPackage(String fullPkgPath) {
+        String orgName;
+        String packageName;
+        String version;
+        // Get org-name
+        int orgNameIndex = fullPkgPath.indexOf("/");
+        if (orgNameIndex == -1) {
+            throw LauncherUtils.createCommandException("no org-name is provided");
```
Comment (MaryamZi) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Following the error templates we decided on, IMO these errors fall into the "command related runtime exceptions category" and should be prefixed with "error: " and not "ballerina: ".

We don't need a new method `LauncherUtils.createCommandException` right? We can use `LauncherUtils.createLauncherException` for the moment, pre-pending "error: " to the error message here. With https://github.com/ballerina-platform/ballerina-lang/pull/10362 we are changing `LauncherUtils.createLauncherException` to prepend "error: ", so you can directly use that method without "error : ".

Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: Fixed with 2546ac7

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    /**
+     * Delete empty parent directories.
+     *
+     * @param from package directory path
+     * @param to   home repository path
+     * @throws IOException throw an exception if an error occurs
+     */
+    private static void deleteEmptyParentDirs(Path from, Path to) throws IOException {
```
Comment (Kishanthan) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/UninstallUtils.java]: I don't feel this method is designed correctly or the params names are misleading. Lets have a chat on this

**Pull request merged**